tests/topology.rs

```diff
@@ -0,0 +1,450 @@
+#[macro_use]
+extern crate tracing;
+
+pub mod support;
+
+use crate::support::{sink, sink_failing_healthcheck, source, transform};
+use futures::{future, future::Future, sink::Sink, stream::iter_ok, stream::Stream, sync::oneshot};
+use std::iter;
+use std::sync::{
+    atomic::{AtomicBool, AtomicUsize, Ordering},
+    Arc,
+};
+use vector::event::{Event, MESSAGE};
+use vector::test_util::{runtime, shutdown_on_idle, trace_init};
+use vector::topology;
+use vector::topology::config::Config;
+
+fn basic_config() -> Config {
+    let mut config = Config::empty();
+    config.add_source("in1", source().1);
+    config.add_sink("out1", &["in1"], sink().1);
+    config
+}
+
+fn basic_config_with_sink_failing_healthcheck() -> Config {
+    let mut config = Config::empty();
+    config.add_source("in1", source().1);
+    config.add_sink("out1", &["in1"], sink_failing_healthcheck().1);
+    config
+}
+
+fn into_message(event: Event) -> String {
+    event.as_log().get(&MESSAGE).unwrap().to_string_lossy()
+}
+
+#[test]
+fn topology_source_and_sink() {
+    let mut rt = runtime();
+    let (in1, source1) = source();
+    let (out1, sink1) = sink();
+
+    let mut config = Config::empty();
+    config.add_source("in1", source1);
+    config.add_sink("out1", &["in1"], sink1);
+
+    let (topology, _crash) = topology::start(config, &mut rt, false).unwrap();
+
+    let event = Event::from("this");
+    let h_out1 = oneshot::spawn(out1.collect(), &rt.executor());
```

Comment:
You don't actually need to spawn this! You can use wait here. The reason is that channels are reactor/timer indepenednt aka they dont do any io. So you can poll them until they finish without an executor/reactor/timer. Before we needed to do this because we were using a TcpStream which needed to be driven by the reactor. Now we can drop this!

So you can do 

```rust
out1.collect().wait().unwrap()
```

This roughly replicates block_on but without needing `oneshot::spawn`.